### ISSUE TITLE
QueryConditions for Dense Arrays

### ISSUE THREAD

**Issue by VictorAlex1** [TileDB-Inc/TileDB-Py#1196]
Hi.

I have following scenario, in which I open a tiledb that contains a Dense Array. I apply QueryConditions on the tiledb obj and it complains that QueryConditions may only be applied to sparse arrays.

```
import numpy as np
import pandas as pd
import tiledb as td

# create the df

columns = ["index", "seqnames", "start", "end"]
records = np.array(
    [
        [0, "chr1", 0, 5],
        [1, "chr1", 6, 10],
        [2, "chr2", 11, 15],
        [3, "chr2", 16, 20],
        [4, "chr3", 21, 25],
    ]
)

df = pd.DataFrame(records, columns=columns)

# write to tiledb

td.from_pandas("./test_pandas", df)


# open tiledb with query

with td.open("./test_pandas", "r") as a:
    qc = td.QueryCondition("seqnames == 'chr1'")
    q = a.query(attr_cond=qc)
```

The traceback:
```
---------------------------------------------------------------------------
TileDBError                               Traceback (most recent call last)
test.py in <cell line: 28>()
     [168](file:///test.py?line=167) with td.open("./test_pandas", "r") as a:
     [169](file:///test.py?line=168)     qc = td.QueryCondition("seqnames == 'chr1'")
---> [170](file:///test.py?line=169)     q = a.query(attr_cond=qc)

File tiledb/libtiledb.pyx:4369, in tiledb.libtiledb.DenseArrayImpl.query()

File tiledb/libtiledb.pyx:4098, in tiledb.libtiledb.Query.__init__()

TileDBError: QueryConditions may only be applied to sparse arrays
```

The documentation implies that this would work with dense arrays. Is there a work around for this?

**Comment by nguyenv** [comment]
Hi @VictorAlex1,

Thanks for reporting this issue. We have just put in a PR to support applying query conditions to dense arrays, and plan to roll out this feature in our next release, 0.16.2.

**Comment by VictorAlex1** [comment]
thank you for the reply :)
i look forward to the release

**Comment by VictorAlexRoche** [comment]
Hi.

I followed the same example on tiledb v 0.16.3 but this time trying to get the dataframe as per:

https://docs.tiledb.com/main/how-to/arrays/reading-arrays/query-conditions

```
import numpy as np
import pandas as pd
import tiledb as td

# create the df

columns = ["index", "seqnames", "start", "end"]
records = np.array(
    [
        [0, "chr1", 0, 5],
        [1, "chr1", 6, 10],
        [2, "chr2", 11, 15],
        [3, "chr2", 16, 20],
        [4, "chr3", 21, 25],
    ]
)

df = pd.DataFrame(records, columns=columns)

# write to tiledb

td.from_pandas("./test_pandas", df)


# open tiledb with query

with td.open("./test_pandas", "r") as a:
    qc = td.QueryCondition("seqnames == 'chr1'")
    q = a.query(attr_cond=qc)
   print(q.df[:])
```

I get the following traceback:

```
---------------------------------------------------------------------------
TileDBError                               Traceback (most recent call last)
/Users/alexanv8/Documents/work/jodi/test.py in <cell line: 7>()
     [118](file:///Users/alexanv8/Documents/work/jodi/test.py?line=117) print(q)
     [119](file:///Users/alexanv8/Documents/work/jodi/test.py?line=118) print(q.df)
---> [120](file:///Users/alexanv8/Documents/work/jodi/test.py?line=119) print(q[:])

File tiledb/libtiledb.pyx:4174, in tiledb.libtiledb.Query.__getitem__()

File tiledb/libtiledb.pyx:4507, in tiledb.libtiledb.DenseArrayImpl.subarray()

File tiledb/libtiledb.pyx:4536, in tiledb.libtiledb.DenseArrayImpl._read_dense_subarray()

TileDBError: [TileDB::QueryCondition] Error: Value node non-empty attribute may only be var-sized for ASCII strings: seqnames
```

And if my query is as following:

```
with td.open("./test_pandas", "r") as a:

    # qc = td.QueryCondition("seqnames == 'chr1'")
    qc = td.QueryCondition("start < 100")
    q = a.query(attr_cond=qc)

    print(q)
    print(q.df)
    print(q[:])
```

I get this traceback:

```
<tiledb.libtiledb.Query object at 0x7fb16dbf6160>
<tiledb.multirange_indexing.DataFrameIndexer object at 0x7fb16d1e6130>
/Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/tiledb/dataframe_.py:39: UserWarning: PyArrow version >= 1.0 is suggested for dataframe functionality.
                  Please `pip install pyarrow>=1.0`.
  warnings.warn(pa_error)
Output exceeds the [size limit](command:workbench.action.openSettings?[). Open the full output data [in a text editor](command:workbench.action.openLargeOutput?3e126f27-5b7a-4499-b372-f88ffe88b17f)
---------------------------------------------------------------------------
TileDBError                               Traceback (most recent call last)
File tiledb/libtiledb.pyx:4532, in tiledb.libtiledb.DenseArrayImpl._read_dense_subarray()

TileDBError: TypeError: init_string(): incompatible function arguments. The following argument types are supported:
    1. (self: tiledb.main.PyQueryCondition, arg0: str, arg1: str, arg2: tiledb_query_condition_op_t) -> None

Invoked with: <tiledb.main.PyQueryCondition object at 0x7fb16c78c030>, 'start', 100, <tiledb_query_condition_op_t.TILEDB_LT: 0>

At:
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/tiledb/query_condition.py(188): aux_visit_Compare
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/tiledb/query_condition.py(155): visit_Compare
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/ast.py(407): visit
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/tiledb/query_condition.py(109): init_query_condition
  <ipython-input-4-36461831f9c9>(15): <cell line: 7>
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/IPython/core/interactiveshell.py(3397): run_code
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/IPython/core/interactiveshell.py(3337): run_ast_nodes
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/IPython/core/interactiveshell.py(3134): run_cell_async
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/IPython/core/async_helpers.py(129): _pseudo_sync_runner
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/IPython/core/interactiveshell.py(2935): _run_cell
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/IPython/core/interactiveshell.py(2880): run_cell
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/ipykernel/zmqshell.py(528): run_cell
  /var/folders/20/y7647_s957dgrrryc9j96ww00000gn/T/ipykernel_10791/740965172.py(28): wrapper
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/ipykernel/ipkernel.py(383): do_execute
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/ipykernel/kernelbase.py(724): execute_request
...
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/traitlets/config/application.py(972): launch_instance
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/site-packages/ipykernel_launcher.py(17): <module>
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/runpy.py(87): _run_code
  /Users/alexanv8/miniconda3/envs/jodi/lib/python3.9/runpy.py(197): _run_module_as_main
 (tiledb/core.cc:714)
```

In this traceback, I dont even know what the error is.

Is there a way to extract a dataframe or ordered dictionary from the Query object?

**Comment by nguyenv** [comment]
The issue is that `from_pandas` in your code is creating an array where all the resulting attributes are unicode type and not the dtypes you are expecting.

ASCII is the only type of string that works with `QueryConditions`. I've set the other attributes to `int` but any NumPy dtype works such as `np.uint8`, etc. It can be set in `from_pandas` like so:

```
import numpy as np
import pandas as pd
import tiledb as td

# create the df
columns = ["index", "seqnames", "start", "end"]
records = np.array(
    [
        [0, "chr1", 0, 5],
        [1, "chr1", 6, 10],
        [2, "chr2", 11, 15],
        [3, "chr2", 16, 20],
        [4, "chr3", 21, 25],
    ]
)

df = pd.DataFrame(records, columns=columns)

# write to tiledb
td.from_pandas(
    "./test_pandas",
    df,
    column_types={"index": int, "start": int, "seqnames": "ascii", "end": int},
)

with td.open("./test_pandas", "r") as a:
    qc = td.QueryCondition("seqnames == 'chr1' and start < 100")
    q = a.query(attr_cond=qc)

    print(q.df[:])
    print(q[:])
```

Output:
```
                 index seqnames                start                  end
0                    0     chr1                    0                    5
1                    1     chr1                    6                   10
2 -9223372036854775808         -9223372036854775808 -9223372036854775808
3 -9223372036854775808         -9223372036854775808 -9223372036854775808
4 -9223372036854775808         -9223372036854775808 -9223372036854775808
OrderedDict([('index', array([                   0,                    1, -9223372036854775808,
       -9223372036854775808, -9223372036854775808])), ('seqnames', array([b'chr1', b'chr1', b'', b'', b''], dtype=object)), ('start', array([                   0,                    6, -9223372036854775808,
       -9223372036854775808, -9223372036854775808])), ('end', array([                   5,                   10, -9223372036854775808,
       -9223372036854775808, -9223372036854775808]))])
```

**Comment by VictorAlexRoche** [comment]
Hi. Thank you for your reply.

I followed your instructions for column_types and QueryCondition works.

However, q.df[:] and q[:] seem to return the negative integer and empty strings (entire df). Is there a way to extract, in this case, only the index column of what the query satisfies? I try to avoid reading the entire df into memory. I looked into the documentation and I couldn't find an example.

**Comment by nguyenv** [comment]
Dense arrays need to return the entire subarray you've requested and fill in non-satisfiable values with defaulted values. You can do some processing afterwards to retrieve only what you need:
```
with td.open("./test_pandas", "r") as a:
    qc = td.QueryCondition("seqnames == 'chr1' and start < 100")
    q = a.query(attr_cond=qc)

    # get default value for attribute
    mask = a.attr("index").fill[0]

    data = q.df[:]

    # filter out the non-satisfied values
    print(data[data["index"] != mask])
```

Output:
```
   index seqnames  start  end
0      0     chr1      0    5
1      1     chr1      6   10
```

**Comment by VictorAlexRoche** [comment]
Hi.

The data that I have generated has over 1.5 mil rows. The following lines take this amount of time:

Apply query:
`q = a.query(attr_cond=qc)`
0.0015 s

Read all into DF:
`data = q.df[:]`
0.18 s

I was wondering if the "application of query" has knowledge of which indexes have been filled with "TileDB default fill value" or if the query condition is applied at "read into df" step.

I try grabbing the indexes of the rows that satisfy the QueryCondition the fastest way possible.

I tried the following scenario where, I thought I could retrieve only the column "start" and the queryCondition has been met:
`q = a.query(attrs=("start",), attr_cond=qc)`
However it complains that the other columns do not exist:
```
TileDBError: TileDBError: Attribute `seqnames` given to filter in query's `attr_cond` arg but not found in `attr` arg.
```

**Comment by stavrospapadopoulos** [comment]
Yeah, I think we should relax the constraint that the condition attribute should be in the `attrs` list. cc @KiterLuc @ihnorton 

> I was wondering if the "application of query" has knowledge of which indexes have been filled with "TileDB default fill value" or if the query condition is applied at "read into df" step.

I am not sure I understood this, can you please elaborate? The query condition is applied before the result buffers are populated and, therefore, before the results are materialized into the pandas dataframe.

**Comment by VictorAlex1** [comment]
@stavrospapadopoulos 
I am trying not to read the entire data frame into memory, especially the data that does not fit the query.

(a.query) looks like the application of the query
(data = q.df[:]) looks like a slicing operation that keeps the data and the fill data together. I was wondering if q at this point has knowledge of the indexes where data exist.

df[:] in particular is expensive whether it’s on a or q . Slicing a single whole column is a lot faster.

**Comment by jkanche** [comment]
@stavrospapadopoulos currently when we use query condition on a dense array that has a million rows, the python interface fills filtered rows with large random negative numbers.

What would be nice is if we can get the indices of rows that match a particular condition. Couple of reasons is to reuse the indices that match a query condition (we can cache these on our side), and choose what column slices I can extract instead of 

- Step 1: scan through the entire file, run the query condition, keep track of which rows match. 
- Step 2: read the Dense Array for matched rows and return. 

What @VictorAlex1 is asking is if we can get the indices of rows that match in Step 1

**Comment by stavrospapadopoulos** [comment]
Hi @jkanche and @VictorAlex1 , 

Yeap, I understand now. Currently we replace values that don't qualify with the special fill values of each attribute (which are configurable at array creation). We are soon adding two more modes:

* Return the result as a sparse array (i.e., in the coordinate format, similar to what we do for sparse arrays - this is what you guys are asking). This is definitely viable, but will incur potentially massive overhead if the number of results is huge.
* Return the result as a dense array, but add a bitmap vector which has 1 in the position of a qualifying cell and 0 for non-qualifying cells. There are applications where this is preferable to the coordinate format.

I hope this helps. Both the above are in our roadmap (cc @ihnorton @KiterLuc ).

**Comment by zegonix** [comment]
> Dense arrays return the entire subarray you've requested and fill in non-satisfiable values with defaulted values.

@nguyenv from what I learned through testing and reading both in the documentation and here, it seems that tiledb takes the smallest subarray that contains all cells which satisfy the first condition. In the example of this issue that would be `seqnames == 'chr1'`. Then all cells which do not satisfy the following conditions, `start < 100` in this issue, are just blanked.

Why is that? I would expect from a query function to just return the cells that satisfy my conditions..
Also, I don't want a pandas frame as output, but a dict, same as what I input. I figured out, that I can get a dict with `<query_object>.multi_index`, but same problem as described before. So WHY? If I need to clean up the output anyways, why even use a query function?
